# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [5]:
#Libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor,AdaBoostRegressor, GradientBoostingRegressor

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [3]:
# Feature Scaling/Selection
# Drop rows containing missing values
spaceship = spaceship.dropna()

# Cabin is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}
spaceship['Cabin'] = spaceship['Cabin'].apply(lambda x: x[0])

# Drop PassengerId and Name
spaceship = spaceship.drop(['PassengerId', 'Name'], axis=1)

# For non-numerical columns, do dummies.
# Asegurarse de que las columnas booleanas estén correctamente tipificadas como bool
spaceship['Transported'] = spaceship['Transported'].astype(bool)
# Seleccionar solo columnas no numéricas
non_numeric_columns = spaceship.select_dtypes(include=['object']).columns

# Convertir columnas no numéricas en dummies
spaceship = pd.get_dummies(spaceship, columns=non_numeric_columns)



**Perform Train Test Split**

In [7]:
# Train Test Split
features = spaceship.drop(columns=['Transported'])
target = spaceship['Transported']

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)

normalizer = MinMaxScaler()

normalizer.fit(X_train)

X_train_norm = normalizer.transform(X_train)

X_test_norm = normalizer.transform(X_test)

In [8]:
X_train_norm = pd.DataFrame(X_train_norm, columns = X_train.columns)
X_train_norm.head()

Age  RoomService  FoodCourt  ShoppingMall       Spa    VRDeck  \
0  0.316456          0.0   0.056116      0.000000  0.028650  0.030094   
1  0.455696          0.0   0.088015      0.135232  0.124911  0.000049   
2  0.430380          0.0   0.000000      0.000000  0.000000  0.000000   
3  0.468354          0.0   0.000000      0.000000  0.000000  0.000000   
4  0.278481          0.0   0.000000      0.000000  0.000000  0.000000   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  ...  \
0               0.0                1.0              0.0              1.0  ...   
1               0.0                1.0              0.0              1.0  ...   
2               0.0                0.0              1.0              0.0  ...   
3               1.0                0.0              0.0              0.0  ...   
4               1.0                0.0              0.0              0.0  ...   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      0.0      0.0      0.0      0.0      0.0                      1.0   
1      0.0      0.0      0.0      0.0      0.0                      0.0   
2      0.0      0.0      1.0      0.0      0.0                      1.0   
3      0.0      0.0      0.0      1.0      0.0                      1.0   
4      0.0      0.0      0.0      1.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                        0.0                      0.0        1.0       0.0  
1                        0.0                      1.0        1.0       0.0  
2                        0.0                      0.0        1.0       0.0  
3                        0.0                      0.0        1.0       0.0  
4                        1.0                      0.0        1.0       0.0  

[5 rows x 24 columns]

In [9]:
X_test_norm = pd.DataFrame(X_test_norm, columns = X_test.columns)
X_test_norm.head()

Age  RoomService  FoodCourt  ShoppingMall       Spa    VRDeck  \
0  0.367089     0.000000   0.000000      0.000000  0.000000  0.000000   
1  0.164557     0.000000   0.000000      0.000000  0.000000  0.000000   
2  0.632911     0.000000   0.000000      0.000000  0.000000  0.000000   
3  0.075949     0.000000   0.000000      0.000000  0.000000  0.000000   
4  0.468354     0.024458   0.000101      0.049049  0.000045  0.003344   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  ...  \
0               0.0                1.0              0.0              0.0  ...   
1               1.0                0.0              0.0              0.0  ...   
2               1.0                0.0              0.0              0.0  ...   
3               1.0                0.0              0.0              1.0  ...   
4               1.0                0.0              0.0              1.0  ...   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      0.0      0.0      0.0      0.0      0.0                      0.0   
1      0.0      0.0      0.0      1.0      0.0                      1.0   
2      0.0      0.0      0.0      1.0      0.0                      0.0   
3      0.0      0.0      0.0      1.0      0.0                      1.0   
4      0.0      0.0      0.0      1.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                        0.0                      1.0        1.0       0.0  
1                        0.0                      0.0        1.0       0.0  
2                        0.0                      1.0        1.0       0.0  
3                        0.0                      0.0        1.0       0.0  
4                        0.0                      1.0        1.0       0.0  

[5 rows x 24 columns]

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [10]:
# Bagging and Pasting
bagging_reg = BaggingRegressor(DecisionTreeRegressor(max_depth=20),
                               n_estimators=100,
                               max_samples = 1000)

bagging_reg.fit(X_train_norm, y_train)

pred = bagging_reg.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", bagging_reg.score(X_test_norm, y_test))

MAE 0.2684000762393858
RMSE 0.3672228114637512
R2 score 0.4605106027879615


- Random Forests

In [11]:
# Random Forests
forest = RandomForestRegressor(n_estimators=100,
                             max_depth=20)

forest.fit(X_train_norm, y_train)

pred = forest.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", forest.score(X_test_norm, y_test))

MAE 0.259882248915011
RMSE 0.37560578373196274
R2 score 0.4355985076207445


- Gradient Boosting

In [12]:
# Gradient Boosting
gb_reg = GradientBoostingRegressor(max_depth=20,
                                   n_estimators=100)

gb_reg.fit(X_train_norm, y_train)

pred = gb_reg.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", gb_reg.score(X_test_norm, y_test))

MAE 0.25099426253820456
RMSE 0.4309313094642839
R2 score 0.25708400407830856


- Adaptive Boosting

In [13]:
# Adaptive Boosting
ada_reg = AdaBoostRegressor(DecisionTreeRegressor(max_depth=20),
                            n_estimators=100)

ada_reg.fit(X_train_norm, y_train)

pred = ada_reg.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", ada_reg.score(X_test_norm, y_test))

MAE 0.2356369825551786
RMSE 0.41721716821899807
R2 score 0.30361733233989185


Which model is the best and why?

#### Bagging and Pasting da los mejores resultados porque da un error absoluto más bajo, un error al cuadrado más bajo que el resto también y sobre todo un R cuadrado superior a los demás, aunque sigue siendo un índice bastante bajo, porque es de 0.46.